src/methods/join.js

```diff
@@ -806,11 +826,26 @@ export class JoinImplementation {
           d2 = b.descriptor; //Get value of property before the split
         }
       }
-      return this.joinDescriptors(realm, joinCondition, d1, d2);
+      be1.descriptor = d1;
+      be2.descriptor = d2;
+      return this.joinBindingEntries(realm, joinCondition, be1, be2);
     };
     return this.joinMaps(m1, m2, join);
   }
 
+  joinBindingEntries(
+    realm: Realm,
+    joinCondition: AbstractValue,
+    be1: PropertyBindingEntry,
+    be2: PropertyBindingEntry
+  ): PropertyBindingEntry {
+    return {
+      descriptor: this.joinDescriptors(realm, joinCondition, be1.descriptor, be2.descriptor),
+      isDeleted: false,
+      previousDescriptor: be1.previousDescriptor === be2.previousDescriptor ? be1.previousDescriptor : undefined,
```

Comment:
I think this line is the cause for the failure of the repro. Can this not be?

`this.joinDescriptors(realm, joinCondition, be1.previousDescriptor, be2.previousDescriptor)`?

Reply:
No, this should be the state before the fork, so joining the descriptor is not appropriate.

Reply:
Either way, we should never be setting the descriptor to `undefined` if both previous descriptors are defined. Otherwise, when we undo the bindings and switch to the previous descriptor, which is undefined, we run into issues like in the failing test case Dan gave above.

Reply:
Agreed. Going to undefined seems like a bug.

Reply:
BTW. even if I apply your suggested change, test-react still fails for me.

Reply:
@hermanventer with master rebased on your PR? Last time I checked it all passed for me.

Reply:
Yes, master with the PR.

Reply:
Fixed in https://github.com/facebook/prepack/pull/2149